File: backend_samples/library/python/AzureAIVisionFaceDeviceAttestation/azure_ai_vision_face_deviceattestation/android/revocation.py
"""
Google Android Key Attestation certificate revocation list:
https://android.googleapis.com/attestation/status

Fetched lazily, cached in-process (Cache-Control max-age honored, default 1h),
fails closed: an unreachable list is treated as REVOKED for every cert.
"""

from __future__ import annotations

import logging
import re
import time
from dataclasses import dataclass, field
from typing import Any, Optional

import requests
from cryptography import x509

from ..logging import track_dependency, track_event, track_exception

log = logging.getLogger("android.revocation")

_REVOCATION_URL = "https://android.googleapis.com/attestation/status"
_DEFAULT_TTL_MS = 3600 * 1000  # 1 hour

_cache: dict[str, Any] = {"data": None, "fetched_at": 0.0, "ttl_ms": _DEFAULT_TTL_MS}


def fetch_revocation_status_list() -> Optional[dict]:
    now = time.time() * 1000
    if _cache["data"] is not None and (now - _cache["fetched_at"]) < _cache["ttl_ms"]:
        track_event(
            "AndroidAuth.RevocationList.CacheHit",
            {
                "ageMs": int(now - _cache["fetched_at"]),
                "ttlMs": _cache["ttl_ms"],
                "entryCount": len((_cache["data"] or {}).get("entries") or {}),
            },
        )
        return _cache["data"]

    dep_start = time.time() * 1000
    try:
        resp = requests.get(_REVOCATION_URL, timeout=10)
        dep_duration = time.time() * 1000 - dep_start
        if not resp.ok:
            log.error("Revocation fetch failed: %s %s", resp.status_code, resp.reason)
            track_dependency(
                name="AndroidAttestation.RevocationList",
                target="android.googleapis.com",
                data=_REVOCATION_URL,
                duration=dep_duration,
                success=False,
                result_code=resp.status_code,
                properties={"statusText": resp.reason},
            )
            return None

        cache_control = resp.headers.get("cache-control") or ""
        m = re.search(r"max-age=(\d+)", cache_control)
        if m:
            _cache["ttl_ms"] = int(m.group(1)) * 1000

        status_list = resp.json()
        _cache["data"] = status_list
        _cache["fetched_at"] = now
        entry_count = len((status_list or {}).get("entries") or {})
        track_dependency(
            name="AndroidAttestation.RevocationList",
            target="android.googleapis.com",
            data=_REVOCATION_URL,
            duration=dep_duration,
            success=True,
            result_code=200,
            properties={"entryCount": entry_count, "ttlMs": _cache["ttl_ms"]},
        )
        return status_list
    except Exception as err:
        dep_duration = time.time() * 1000 - dep_start
        track_dependency(
            name="AndroidAttestation.RevocationList",
            target="android.googleapis.com",
            data=_REVOCATION_URL,
            duration=dep_duration,
            success=False,
            result_code="exception",
        )
        track_exception(err, {"source": "fetch_revocation_status_list"})
        return None


def check_certificate_revocation(cert_der: bytes, status_list: Optional[dict]) -> dict:
    """Return {isRevoked, reason?, status?}."""
    if not status_list or not status_list.get("entries"):
        # Fail closed
        log.error("Revocation list unavailable — failing closed")
        return {"isRevoked": True, "reason": "REVOCATION_CHECK_UNAVAILABLE", "status": "REVOKED"}

    try:
        cert = x509.load_der_x509_certificate(cert_der)
        serial_hex = format(cert.serial_number, "x").lower()
        entry = status_list["entries"].get(serial_hex)
        if entry:
            return {
                "isRevoked": True,
                "status": entry.get("status"),
                "reason": entry.get("reason") or "UNSPECIFIED",
            }
        return {"isRevoked": False}
    except Exception as err:
        log.error("Revocation check error: %s", err)
        return {"isRevoked": False}

File: backend_samples/library/java/AzureAIVisionFaceDeviceAttestation/src/main/java/com/azure/ai/vision/face/deviceattestation/android/Revocation.java
package com.azure.ai.vision.face.deviceattestation.android;

import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.security.cert.X509Certificate;
import java.time.Duration;
import java.time.Instant;
import java.util.Map;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import com.azure.ai.vision.face.deviceattestation.Json;
import com.azure.ai.vision.face.deviceattestation.JsonData;
import com.azure.ai.vision.face.deviceattestation.Maps;
import com.azure.ai.vision.face.deviceattestation.crypto.CertUtils;
import com.azure.ai.vision.face.deviceattestation.logging.AttestationLogger;
import com.azure.ai.vision.face.deviceattestation.logging.DependencyTelemetry;

/**
 * Google Android Key Attestation certificate revocation list, fetched lazily
 * and cached in-process. {@link #checkCertificateRevocation} fails closed: if
 * the list can't be fetched, every cert is treated as revoked.
 * https://android.googleapis.com/attestation/status
 */
public final class Revocation {

    private static final String REVOCATION_STATUS_URL = "https://android.googleapis.com/attestation/status";
    private static final Pattern MAX_AGE = Pattern.compile("max-age=(\\d+)");

    private static final HttpClient HTTP = HttpClient.newHttpClient();
    private static final Object CACHE_LOCK = new Object();

    private static StatusList cache;
    private static Instant fetchedAt;
    private static Duration ttl = Duration.ofHours(1);

    /** The parsed revocation entries (serial-hex -> {status, reason}). */
    public record StatusList(Map<String, Object> entries) {
    }

    /** Whether a certificate is revoked, plus the status/reason when it is. */
    public record RevocationResult(boolean isRevoked, String status, String reason) {
    }

    @SuppressWarnings("unchecked")
    public static StatusList fetchRevocationStatusList(AttestationLogger logger) {
        synchronized (CACHE_LOCK) {
            if (cache != null && Duration.between(fetchedAt, Instant.now()).compareTo(ttl) < 0) {
                return cache;
            }
        }

        long start = System.nanoTime();
        try {
            HttpRequest request = HttpRequest.newBuilder(URI.create(REVOCATION_STATUS_URL)).GET().build();
            HttpResponse<String> response = HTTP.send(request, HttpResponse.BodyHandlers.ofString());
            long durationMs = (System.nanoTime() - start) / 1_000_000;

            if (response.statusCode() != 200) {
                logger.trackDependency(dependency(durationMs, false, String.valueOf(response.statusCode()), 0));
                return null;
            }

            response.headers().firstValue("cache-control").ifPresent(cc -> {
                Matcher m = MAX_AGE.matcher(cc);
                if (m.find()) {
                    ttl = Duration.ofSeconds(Long.parseLong(m.group(1)));
                }
            });

            Map<String, Object> root = Json.parseObject(response.body());
            Map<String, Object> entries = root != null && root.get("entries") instanceof Map
                    ? (Map<String, Object>) root.get("entries")
                    : null;
            StatusList statusList = new StatusList(entries);

            synchronized (CACHE_LOCK) {
                cache = statusList;
                fetchedAt = Instant.now();
            }

            logger.trackDependency(dependency(durationMs, true, "200", entries != null ? entries.size() : 0));
            return statusList;
        } catch (Exception e) {
            long durationMs = (System.nanoTime() - start) / 1_000_000;
            logger.trackDependency(dependency(durationMs, false, "exception", 0));
            logger.trackException(e, Maps.of("source", "fetchRevocationStatusList"));
            return null;
        }
    }

    @SuppressWarnings("unchecked")
    public static RevocationResult checkCertificateRevocation(byte[] certDer, StatusList statusList) {
        if (statusList == null || statusList.entries() == null) {
            // Fail closed: an attestation we can't check against revocations is rejected.
            return new RevocationResult(true, "REVOKED", "REVOCATION_CHECK_UNAVAILABLE");
        }

        try {
            X509Certificate cert = CertUtils.loadCertificate(certDer);
            String serialHex = cert.getSerialNumber().toString(16);
            Object entry = statusList.entries().get(serialHex);
            if (entry instanceof Map) {
                Map<String, Object> e = (Map<String, Object>) entry;
                String reason = JsonData.getString(e, "reason");
                return new RevocationResult(true, JsonData.getString(e, "status"), reason != null ? reason : "UNSPECIFIED");
            }
            return new RevocationResult(false, null, null);
        } catch (Exception e) {
            return new RevocationResult(false, null, null);
        }
    }

    private static DependencyTelemetry dependency(long durationMs, boolean success, String resultCode, int entryCount) {
        return DependencyTelemetry.builder("AndroidAttestation.RevocationList")
                .target("android.googleapis.com")
                .data(REVOCATION_STATUS_URL)
                .durationMillis(durationMs)
                .success(success)
                .resultCode(resultCode)
                .properties(success ? Maps.of("entryCount", entryCount) : null)
                .build();
    }

    private Revocation() {
    }
}
